cpio: buggy upstream patch in nixpkgs master

Practical consequence: `pkgsi686Linux.linux` [fails](https://hydra.nixos.org/build/154224895#tabs-summary) with
```
  CHK     kernel/kheaders_data.tar.xz
  GEN     kernel/kheaders_data.tar.xz
cpio: t.h: Cannot stat: No such file or directory
cpio: tor.h: Cannot stat: No such file or directory
cpio: /s2mpu02.h: Cannot stat: No such file or directory
cpio: /brcmstb.h: Cannot stat: No such file or directory
cpio: bflush.h: Cannot stat: No such file or directory
cpio: estart.h: Cannot stat: No such file or directory
cpio: _types.h: Cannot stat: No such file or directory
cpio: nfnetlink_log.h: Cannot stat: No such file or directory
cpio: _ipt.h: Cannot stat: No such file or directory
cpio: blank line ignored
make[2]: *** [../kernel/Makefile:151: kernel/kheaders_data.tar.xz] Error 2
make[1]: *** [/build/linux-5.10.68/Makefile:1822: kernel] Error 2
make: *** [../Makefile:185: __sub-make] Error 2
```
(this log snippet is cleaner thanks to `--cores 1`)
I checked that it really is the patch by applying
```diff
--- a/pkgs/os-specific/linux/kernel/manual-config.nix
+++ b/pkgs/os-specific/linux/kernel/manual-config.nix
@@ -316,7 +316,10 @@ stdenv.mkDerivation ((drvAttrs config stdenv.hostPlatform.linux-kernel kernelPat
       ++ optional  (lib.versionAtLeast version "4.14" && lib.versionOlder version "5.8") libelf
       # Removed util-linuxMinimal since it should not be a dependency.
       ++ optionals (lib.versionAtLeast version "4.16") [ bison flex ]
-      ++ optionals (lib.versionAtLeast version "5.2")  [ cpio pahole ]
+      ++ optionals (lib.versionAtLeast version "5.2")  [
+            (cpio.overrideAttrs (_: { patches = null; }))
+            pahole
+          ]
       ++ optional  (lib.versionAtLeast version "5.8")  elfutils
       ;
 
```

Some bugs are discussed in the upstream ML thread: https://lists.gnu.org/archive/html/bug-cpio/2021-08/msg00000.html

I'm not sure about the best approach for us so far; the rebuild amounts complicate this.  Perhaps just temporarily fixup the kernel build and get a properly fixed cpio with the next `staging-next` iteration?

I expect that three to five upstream commits will be needed; maybe just the whole diff since the last release.  https://git.savannah.gnu.org/cgit/cpio.git/log/

--- a/pkgs/tools/archivers/cpio/default.nix
+++ b/pkgs/tools/archivers/cpio/default.nix
@@ -1,13 +1,11 @@
 { lib, stdenv, fetchurl, fetchpatch }:
 
-let
+stdenv.mkDerivation rec {
+  pname = "cpio";
   version = "2.13";
-  name = "cpio-${version}";
-in stdenv.mkDerivation {
-  inherit name;
 
   src = fetchurl {
-    url = "mirror://gnu/cpio/${name}.tar.bz2";
+    url = "mirror://gnu/cpio/cpio-${version}.tar.bz2";
     sha256 = "0vbgnhkawdllgnkdn6zn1f56fczwk0518krakz2qbwhxmv2vvdga";
   };
 
@@ -26,9 +24,9 @@ in stdenv.mkDerivation {
         "0pidkbxalpj5yz4fr95x8h0rizgjij0xgvjgirfkjk460giawwg6")
   ];
 
-  preConfigure = if stdenv.isCygwin then ''
+  preConfigure = lib.optionalString stdenv.isCygwin ''
     sed -i gnu/fpending.h -e 's,include <stdio_ext.h>,,'
-  '' else null;
+  '';
 
   enableParallelBuilding = true;
 

--- a/pkgs/tools/archivers/cpio/default.nix
+++ b/pkgs/tools/archivers/cpio/default.nix
@@ -11,12 +11,19 @@ in stdenv.mkDerivation {
     sha256 = "0vbgnhkawdllgnkdn6zn1f56fczwk0518krakz2qbwhxmv2vvdga";
   };
 
-  patches = [
-    (fetchpatch {
-      name = "CVE-2021-38185.patch";
-      url = "https://git.savannah.gnu.org/cgit/cpio.git/patch/?id=dd96882877721703e19272fe25034560b794061b";
-      sha256 = "0vmr0qjwj2ldnzsvccl105ckwgx3ssvn9mp3f27ss0kiyigrzz32";
-    })
+  patches = let
+    fp = suffix: rev: sha256: fetchpatch {
+      name = "CVE-2021-38185-${suffix}.patch";
+      url = "https://git.savannah.gnu.org/cgit/cpio.git/patch/?id=${rev}";
+      inherit sha256;
+    };
+  in [
+    (fp "1" "dd96882877721703e19272fe25034560b794061b"
+        "0vmr0qjwj2ldnzsvccl105ckwgx3ssvn9mp3f27ss0kiyigrzz32")
+    (fp "2" "dfc801c44a93bed7b3951905b188823d6a0432c8"
+        "1qkrhi3lbxk6hflp6w3h4sgssc0wblv8r0qgxqzbjrm36pqwxiwh")
+    (fp "3" "236684f6deb3178043fe72a8e2faca538fa2aae1"
+        "0pidkbxalpj5yz4fr95x8h0rizgjij0xgvjgirfkjk460giawwg6")
   ];
 
   preConfigure = if stdenv.isCygwin then ''

